### ISSUE TITLE
Question: possible to add a custom hook to get session information on master when using xdist?

### ISSUE THREAD

**Issue by symonk** [pytest-dev/pytest#6325]
Hi, I thought this was better suited here instead of the xdist repo.  Since using xdist there seems to be no easy way of getting session items on the master process, is it possible to add my own hook to do this?

I need to get the collection of items to be executed because I need to do something based on that information on master ONLY, not a gateway node. 

How can I achieve this? I know I can register my own hooks but how do I tell them where they should run and which information they can receive etc?

Master doesn’t seem to run the collection spec hooks , seems to only hit configure and unconfigure?

I need collection to have happened and then have the information on the master mode, as I want to do logic based on fixturenames and iter_markers.

TLDR: I need access to which markers are fixtures will be executed in the run on the master process, i want to spin up some heavy server IF a marker or fixture is to be used at runtime, i don't want to modify the items just dictate if the criteria will be met and do something standalone

Thank you for your time.

**Comment by nicoddemus** [comment]
Hi @symonk,

You can implement `pytest_collection` yourself in a `conftest.py` file:

```python
def pytest_collection(session):
    items = session.perform_collect()
    ... # do something with the items, analyse markers and fixtures
    return items
```

**Comment by symonk** [comment]
@nicoddemus thanks, will this run on the master node tho?

**Comment by nicoddemus** [comment]
Strange, debugging it here I can see that it is calling the hooks from the master worker:

![image](https://user-images.githubusercontent.com/1085180/70391311-900f0d00-19b2-11ea-8a2b-de16ba7799bd.png)

(`pytest_collection` is being called from the custom `conftest.py`)

Compare this when the code is executing in a worker (the bottom of the traceback starts in `execnet` code):

![image](https://user-images.githubusercontent.com/1085180/70391323-b0d76280-19b2-11ea-8571-35d3915f5cb1.png)


If I also add this to my `conftest.py` file:

```python
def pytest_collection_modifyitems(items):
    with open('out.txt', 'a') as f:
        print(f"hey: {os.environ.get('PYTEST_XDIST_WORKER')} {len(items)}", file=f)
```

Then `out.txt` file contains:

```
hey: None 1
hey: gw0 1
```

(`hey:None` means the master worker)

So things seem to be working. Why do you say the usual collection hooks are not being called from the master worker? Can you provide a reproducible example?

**Comment by symonk** [comment]
@nicoddemus this is my question :) perhaps master runs pytest_collection() which would solve my problem i think, i didnt see this hook, but does it run modifyitems/after collection hooks? it is not the behaviour i see. 

I was just trying with:

```
def pytest_collection_modifyitems(session, config, items):
    """ called after collection has been performed, may filter or re-order
    the items in-place.

    :param _pytest.main.Session session: the pytest session object
    :param _pytest.config.Config config: pytest config object
    :param List[_pytest.nodes.Item] items: list of item objects
    """



[docs]def pytest_collection_finish(session):
    """ called after collection has been performed and modified.

    :param _pytest.main.Session session: the pytest session object
    """
```

without implementing a pytest_collection() I guess?

**Comment by nicoddemus** [comment]
I'm confused, did you try my suggestion in https://github.com/pytest-dev/pytest/issues/6325#issuecomment-562951341?

**Comment by symonk** [comment]
@nicoddemus yes, but i dont have it running pytest_collection(session): - however im doing a check based on this:

os.environ.get('PYTEST_CURRENT_WORKER', 'master') (is this flawed?)

perhaps i should be using:

hasattr(config, 'slaveinput') ?

**Comment by symonk** [comment]
i cant get master to run the pytest_configure(session) call, are you sure it works? i thought master doesnt typically do collection, just ensure worker collections are identicle?

**Comment by symonk** [comment]
```python
def pytest_configure(config):
def pytest_unconfigure(config):
```

Above is working OK, but below is not executing on the master:

```python
def pytest_collection(session):
    if not hasattr(session.config, 'slaveinput'):
        return

    config = session.config
    items = session.items
```

N.b this is in a plugin that i have loaded in the root conftest, i'm at a loss

**Comment by symonk** [comment]
```python
def pytest_collection(session):
    log.info(dir(session.config))
```

This only logs out 3x, gw0, gw1, gw2 when running -n=3.  Master is never executing this hook for my plugin atleast...

do I need to decorate the hook with something?
do I need to stop doing it in a plugin that is loaded using pytest_plugins='myplugin' and do it in a conftest.py instead??

**Comment by nicoddemus** [comment]
Oh you are in a plugin? Then the hook order is unfortunately arbirtrary. Try decorating your `pytest_collection` with `@pytest.hookimpl(tryfirst=True)`

**Comment by symonk** [comment]
@nicoddemus I added the decorator, but master is never entering this hook implementation it seems, hmm I'm really confused

**Comment by symonk** [comment]
err, I think @pytest.mark.tryfirst on the pytest_collection  fixed it, could you explain why?

**Comment by symonk** [comment]
n.b I never returned items after doing session.perform_collect() that is possibly why it didn't work by tryfirst does, but i have no idea what that actually does

**Comment by nicoddemus** [comment]
Good to hear.

`tryfirst` will attempt to execute this hook in front of the others, so this will ensure your hook is called before pytest-xdist's.

More info: https://pluggy.readthedocs.io/en/latest/index.html#call-time-order

I'm closing this for now then given that it seems you have found a solution. 👍
